diff --git a/src/Controller/ImageStyleDownloadController.php b/src/Controller/ImageStyleDownloadController.php
index eb992d3..7efbae8 100644
--- a/src/Controller/ImageStyleDownloadController.php
+++ b/src/Controller/ImageStyleDownloadController.php
@@ -155,7 +155,20 @@ class ImageStyleDownloadController extends FileDownloadController {
     // starts with styles/.
     $valid = !empty($image_style) && $this->streamWrapperManager->isValidScheme($scheme);
     if (!$this->config('image.settings')->get('allow_insecure_derivatives') || strpos(ltrim($target, '\/'), 'styles/') === 0) {
-      $valid &= $request->query->get(IMAGE_DERIVATIVE_TOKEN) === $image_style->getPathToken($image_uri);
+      if ($request->query->get(IMAGE_DERIVATIVE_TOKEN) === $image_style->getPathToken($image_uri)) {
+        $valid &= TRUE;
+      }
+      else {
+        $image_uri_exploded = explode('.', $image_uri);
+        $extension = isset($image_uri_exploded[1]) ? $image_uri_exploded[1] : '';
+        $filename = isset($image_uri_exploded[0]) ? $image_uri_exploded[0] : $image_uri;
+        if ($request->query->get(IMAGE_DERIVATIVE_TOKEN) === $image_style->getPathToken($filename . '.' . mb_strtolower($extension))) {
+          $valid &= TRUE;
+        }
+        else {
+          $valid &= FALSE;
+        }
+      }
     }
     if (!$valid) {
       throw new AccessDeniedHttpException();
