diff --git a/src/LayoutBuilderMigration.php b/src/LayoutBuilderMigration.php
index f1fd2a44b..19298056b 100644
--- a/src/LayoutBuilderMigration.php
+++ b/src/LayoutBuilderMigration.php
@@ -215,6 +215,17 @@ private function toSection(array &$configuration, $entity_type_id, $bundle) {
           $block['context_mapping']['entity'] = 'layout_builder.entity';
         }
       }
+      else if ($block['provider'] === 'ctools' && strpos($block['id'], 'entity_view:') === 0) {
+        $block['provider'] = 'layout_builder';
+        // Remove configuration keys that are moved to component-level settings.
+        unset($block['formatter']['region'], $block['formatter']['weight']);
+
+        // If the entity being panelized is referenced in the context mapping,
+        // use the Layout Builder version of that.
+        if (isset($block['context_mapping']['entity']) && $block['context_mapping']['entity'] === '@panelizer.entity_context:entity') {
+          $block['context_mapping']['entity'] = 'layout_builder.entity';
+        }
+      }

       $plugin_definition = $this->blockManager->getDefinition($block['id']);
       // The required context values must be passed directly in the plugin
